tensornetwork/quantum/quantum.py

```diff
@@ -0,0 +1,374 @@
+  return quantum_constructor(out_edges, in_edges)
+
+
+def check_spaces(edges_1: Sequence[Edge], edges_2: Sequence[Edge]):
+  """Check the vector spaces represented by two lists of edges are compatible.
+
+  The number of edges must be the same and the dimensions of each pair of edges
+  must match. Otherwise, an exception is raised.
+
+  Args:
+    edges_1: List of edges representing a many-body Hilbert space.
+    edges_2: List of edges representing a many-body Hilbert space.
+  """
+  if len(edges_1) != len(edges_2):
+    raise ValueError("Hilbert-space mismatch: Cannot connect {} subsystems "
+                     "with {} subsystems.".format(len(edges_1), len(edges_2)))
+
+  for (i, (e1, e2)) in enumerate(zip(edges_1, edges_2)):
+    if e1.dimension != e2.dimension:
+      raise ValueError("Hilbert-space mismatch on subsystems {}: Input "
+                       "dimension {} != output dimension {}.".format(
+                           i, e1.dimension, e2.dimension))
+
+
+class QuOperator():
+  """Represents an operator via a tensor network.
+
+  Can be used to do simple linear algebra with tensor networks.
+  """
+  def __init__(self, out_edges: Sequence[Edge], in_edges: Sequence[Edge],
+               ref_nodes: Optional[Collection[Node]] = None,
+               ignore_edges: Optional[Collection[Edge]] = None):
+    # TODO: Decide whether the user must also supply all nodes involved.
+    #       This would enable extra error checking and is probably clearer
+    #       than `ref_nodes`.
+    if len(in_edges) == 0 and len(out_edges) == 0 and not ref_nodes:
+      raise ValueError("At least one reference node is required to specify a "
+                       "scalar. None provided!")
+    self.out_edges = list(out_edges)
+    self.in_edges = list(in_edges)
+    self.ignore_edges = set(ignore_edges) if ignore_edges else set()
+    self.ref_nodes = set(ref_nodes) if ref_nodes else set()
+    self.check_network()
+
+  @classmethod
+  def from_tensor(cls, tensor: Tensor, out_axes: Sequence[int],
+                  in_axes: Sequence[int], backend: Optional[Text] = None):
+    n = Node(tensor, backend=backend)
+    out_edges = [n[i] for i in out_axes]
+    in_edges = [n[i] for i in in_axes]
+    return cls(out_edges, in_edges, set([n]))
+
+  @property
+  def nodes(self):
+    """All tensor-network nodes involved in the operator.
+    """
+    return reachable(
+        get_all_nodes(self.out_edges + self.in_edges) | self.ref_nodes)
+
+  @property
+  def shape_in(self):
+    return [e.dimension for e in self.in_edges]
+
+  @property
+  def shape_out(self):
+    return [e.dimension for e in self.out_edges]
+
+  @property
+  def is_scalar(self):
+    return len(self.out_edges) == 0 and len(self.in_edges) == 0
+
+  @property
+  def is_vector(self):
+    return len(self.out_edges) > 0 and len(self.in_edges) == 0
+
+  @property
+  def is_adjoint_vector(self):
+    return len(self.out_edges) == 0 and len(self.in_edges) > 0
```

Comment:
Okay, I suppose I don't mind. What's your thinking?

Reply:
Also, do you mean make them methods, or really functions (`def is_vector(op):`...)? Am happy with the former, but find the latter odd, since these functions only make sense for `QuOperator`s.

Reply:
Just remove the `@property` and make them member methods.